libswift/CMakeLists.txt

```diff
@@ -0,0 +1,142 @@
+# This source file is part of the Swift.org open source project
+#
+# Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+# Licensed under Apache License v2.0 with Runtime Library Exception
+#
+# See http://swift.org/LICENSE.txt for license information
+# See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+
+function(add_libswift name module_files_var)
+  cmake_parse_arguments(LIBSWIFT
+                        ""
+                        "SWIFT_EXEC;BUILD_DIR"
+                        ""
+                        ${ARGN})
+
+  set(libswiftModules "SIL" "Optimizer")
+  find_program(swiftpm_executable "swift-build")
+
+  # It would be nice if we could use swiftpm for all builds. But this is
+  # currently not possible. Mainly because we cannot assume that a swift
+  # toolchain is installed on the host system.
+  if(swiftpm_executable AND CMAKE_BUILD_TYPE STREQUAL "Debug")
+
+    if(CMAKE_BUILD_TYPE STREQUAL Debug)
+      set(configuration "debug")
+    else()
+      set(configuration "release")
+    endif()
+
+    set(build_dir "${LIBSWIFT_BUILD_DIR}/${configuration}")
+    file(GLOB_RECURSE sources "*.swift" "${CMAKE_SOURCE_DIR}/include/swift/*Bridging.h")
+
+    # Compile the libswift package using swiftpm.
+    #
+    # Explicitly add dependencies, so that the "cmake" ninja will trigger
+    # swiftpm if any of the libswift files changed.
+
+    if(NOT LIBSWIFT_SWIFT_EXEC STREQUAL "swiftc")
+      set(env_opts "env" "SWIFT_EXEC=${LIBSWIFT_SWIFT_EXEC}")
+    endif()
+
+    # Use the installed swift tools.
+    # TODO: maybe not hardcode "swift", but use a CMAKE variable?
+    add_custom_command(OUTPUT "${build_dir}/libSwift.a"
+      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+      DEPENDS ${sources}
+      COMMAND ${CMAKE_COMMAND} "-E" ${env_opts}
+        "swift" "build" "-c" "${configuration}" "--build-path" "${LIBSWIFT_BUILD_DIR}"
+      COMMENT "Building libswift")
+
+    # We need a target to "link" the custom command with the library.
+    add_custom_target("${name}-swiftpm" DEPENDS "${build_dir}/libSwift.a")
+
+    # Import the built library 
+    add_library(${name} STATIC IMPORTED GLOBAL)
+    add_dependencies(${name} "${name}-swiftpm")
+    set_target_properties(${name}
+        PROPERTIES
+        IMPORTED_LOCATION "${build_dir}/libSwift.a")
+  else()
+
+    # Build with cmake commands instead of using swiftpm.
+
+    if(CMAKE_BUILD_TYPE STREQUAL Debug)
+      set(libswift_compile_options "-g")
+    else()
+      set(libswift_compile_options "-O" "-cross-module-optimization")
+    endif()
+
+    set(build_dir ${LIBSWIFT_BUILD_DIR})
+
+    # Due to a cmake bug it's not possible to use cmake's builtin swift support
+    # for building. There is a problem with static swift libraries which are
```

Comment:
I was thinking... in truth this is going to be extra weight that will have to be fixed by someone at some point. I think that we can make the more invasive part of this patch local to libswift. I also would like to make sure that we have a 3.19 implementation that works before this goes in. (We can compare the version). I am happy to do that part.